enhancements/baremetal/continuous-hw-introspection.md

```diff
@@ -0,0 +1,147 @@
+
+This proposal talks about how performing continuous hardware introspection on all master
+and worker nodes in a Baremetal IPI deployment would get us important data about baremetal
+hosts which can be used to monitor changes in hardware capabilities of the server.
+
+This enhancement will not result in a new CRD or API to be written but will detail
+data that would become availabe in the Status field of the BareMetalHost CRD.
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [X] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Continuous hardware inspection on all the master and worker nodes that are a part
+of a Baremetal IPI deployment, allow the user to detect changes in server hardware
+and become aware of any failures seen on the servers. This information can be used
+to report current status of hardware to the user and/or glean metrics about the
+health of the deployment.
+
+At the end of 4.4, a Baremetal IPI deployment does not have continuous hardware introspection
+running on its master or worker nodes. The underlying provisioning service already has the
+capability to perform hardware inspection on each of the servers periodically. The
+baremetal-operator (BMO) then reads this information and updates the "status" field of the
+BareMetalHost CRD. Please refer to [1] for the latest BaremetalHost CRD.
+
+## Motivation
+
+This enhancement will allow the user to get continuous updates on all the baremetal
+servers in their Metal3 deployment providing useful information about the hardware
+capabilities of the servers.
+
+### Goals
+
+This enhancement will allow the baremetal host CR associated with each baremetal
+server in a metal3 deployment to contain up-to-date information about its hardware and
+operational status of a server.
+
+### Non-Goals
+
+This proposal does not include any changes to the BareMetalHost CRD.
+
+## Proposal
+
+This enhancement proposes to add a new DaemonSet started by the machine-api-operator
+only for the BareMetal platform. This DaemonSet would run a new container image
+(ironic-hardware-inventory-recorder) on all master and worker nodes. With the new
+addition, it would become possible to have detailed information regarding the
+hardware connected to both master and worker nodes in a metal3 deployment and not
+just when the worker nodes are provisioned. This also allows the user to keep
+track of any changes in hardware capabilities of servers in a metal3 deployment.
+
+Additional details regarding enhancements made to the underlying service required
+for this feature can be found in [3].
+
+### User Stories [optional]
+
+#### Story 1
+
+As a user, I would like to know the current status of the hardware and its capabilities
```

Comment:
Starting a story with "as a user" has a tendency to blur the lines between what can be quite different roles - especially between people operating applications on top of the cluster and those operating the cluster itself. Sometimes those are the same people, but often they're not.

In this case, the information is of interest to the person whose job it is is to manage the physical hardware through the metal³ resources. But a lot of the comments are interpreting it as primarily being of interest to application developers needing to find ways of determining which nodes to schedule their applications to. Perhaps being more explicit could help to reduce the confusion.

It might also help to be more explicit about some of the use cases. e.g. "As an administrator tasked with managing the physical infrastructure, I'd like to be able to determine the serial number of every disk currently in use and trace them back to a particular host." I suspect that we don't yet have a completely clear idea of what all of these things might be used for (at least I don't) and that might be part of the difficulty with explaining it.

Reply:
Yes, I think this data is useful mostly to a cloud admin and not to an end user who trying to run an application on this infrastructure. Let me try to reword.